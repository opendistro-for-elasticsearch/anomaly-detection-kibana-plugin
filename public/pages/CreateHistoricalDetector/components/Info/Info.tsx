/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { EuiFieldText, EuiTextArea } from '@elastic/eui';
import { Field, FieldProps } from 'formik';
import React from 'react';
import ContentPanel from '../../../../components/ContentPanel/ContentPanel';
import { getError, isInvalid } from '../../../../utils/utils';
import { FormattedFormRow } from '../../../createDetector/components/FormattedFormRow/FormattedFormRow';

interface InfoProps {
  onValidateDetectorName: (detectorName: string) => any;
  onValidateDetectorDescription: (detectorDescription: string) => any;
}
export function Info(props: InfoProps) {
  return (
    <ContentPanel title="Name and description" titleSize="s">
      <Field name="name" validate={props.onValidateDetectorName}>
        {({ field, form }: FieldProps) => (
          <FormattedFormRow
            title="Name"
            hint="Specify a unique and descriptive name that is easy to
          recognize. Detector name must contain 1-64 characters. 
          Valid characters are a-z, A-Z, 0-9, -(hyphen) and _(underscore)"
            isInvalid={isInvalid(field.name, form)}
            error={getError(field.name, form)}
          >
            <EuiFieldText
              name="name"
              id="name"
              placeholder="Enter detector name"
              isInvalid={isInvalid(field.name, form)}
              onBlur={() => {
                form.setFieldTouched('name', true);
              }}
              {...field}
            />
          </FormattedFormRow>
        )}
      </Field>
      <Field name="description" validate={props.onValidateDetectorDescription}>
        {({ field, form }: FieldProps) => (
          <FormattedFormRow
            formattedTitle={
              <p>
                Description <span className="optional">- optional</span>
              </p>
            }
            hint="Describe the purpose of the detector."
            isInvalid={isInvalid(field.name, form)}
            error={getError(field.name, form)}
          >
            <EuiTextArea
              name="description"
              id="description"
              rows={3}
              placeholder="Enter detector description"
              isInvalid={isInvalid(field.name, form)}
              onBlur={() => {
                form.setFieldTouched('description', true);
              }}
              {...field}
            />
          </FormattedFormRow>
        )}
      </Field>
    </ContentPanel>
  );
}
