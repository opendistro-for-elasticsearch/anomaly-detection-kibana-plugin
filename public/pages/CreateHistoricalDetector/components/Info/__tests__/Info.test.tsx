/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Info } from '../Info';
import { Formik } from 'formik';

const MOCK_INVALID_NAME = 'Invalid detector name';
const MOCK_INVALID_DESCRIPTION = 'Invalid detector description';

const mockValidateDescription = jest.fn().mockImplementation(() => {
  return MOCK_INVALID_DESCRIPTION;
});
const mockValidateName = jest.fn().mockImplementation(() => {
  return MOCK_INVALID_NAME;
});

describe('<Info /> spec', () => {
  test('renders the component', () => {
    const { container } = render(
      <Formik initialValues={{ detectorName: '' }} onSubmit={jest.fn()}>
        {() => (
          <div>
            <Info
              onValidateDetectorName={jest.fn()}
              onValidateDetectorDescription={jest.fn()}
            />
          </div>
        )}
      </Formik>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  test('shows error for invalid detector name input', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik
        initialValues={{ detectorName: '', detectorDescription: '' }}
        onSubmit={jest.fn()}
      >
        {() => (
          <div>
            <Info
              onValidateDetectorName={mockValidateName}
              onValidateDetectorDescription={jest.fn()}
            />
          </div>
        )}
      </Formik>
    );
    expect(queryByText(MOCK_INVALID_NAME)).toBeNull();
    fireEvent.focus(getByPlaceholderText('Enter detector name'));
    fireEvent.blur(getByPlaceholderText('Enter detector name'));
    expect(findByText(MOCK_INVALID_NAME)).not.toBeNull();
  });
  test('shows error for invalid detector description', async () => {
    const { queryByText, findByText, getByPlaceholderText } = render(
      <Formik
        initialValues={{ detectorName: '', detectorDescription: '' }}
        onSubmit={jest.fn()}
      >
        {() => (
          <div>
            <Info
              onValidateDetectorName={jest.fn()}
              onValidateDetectorDescription={mockValidateDescription}
            />
          </div>
        )}
      </Formik>
    );
    expect(queryByText(MOCK_INVALID_DESCRIPTION)).toBeNull();
    fireEvent.focus(getByPlaceholderText('Enter detector description'));
    fireEvent.blur(getByPlaceholderText('Enter detector description'));
    expect(findByText(MOCK_INVALID_DESCRIPTION)).not.toBeNull();
  });
});
