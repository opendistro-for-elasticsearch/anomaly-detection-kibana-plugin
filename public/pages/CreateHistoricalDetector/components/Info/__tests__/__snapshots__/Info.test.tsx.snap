// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Info /> spec renders the component 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Name and description
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 10px 0px;"
    >
      <div
        class="euiFormRow"
        hint="Specify a unique and descriptive name that is easy to recognize. Detector name must contain 1-64 characters.  Valid characters are a-z, A-Z, 0-9, -(hyphen) and _(underscore)"
        id="random_html_id-row"
        title="Name"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            aria-invalid="false"
            class="euiFormLabel euiFormRow__label"
            for="random_html_id"
          >
            <div
              style="line-height: 8px;"
            >
              <p>
                Name
              </p>
              <br />
              <p
                class="sublabel"
              >
                Specify a unique and descriptive name that is easy to recognize. Detector name must contain 1-64 characters.  Valid characters are a-z, A-Z, 0-9, -(hyphen) and _(underscore)
              </p>
            </div>
          </label>
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        >
          <div
            class="euiFormControlLayout"
          >
            <div
              class="euiFormControlLayout__childrenWrapper"
            >
              <input
                class="euiFieldText"
                id="random_html_id"
                name="name"
                placeholder="Enter detector name"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiFormRow"
        hint="Describe the purpose of the detector."
        id="random_html_id-row"
      >
        <div
          class="euiFormRow__labelWrapper"
        >
          <label
            aria-invalid="false"
            class="euiFormLabel euiFormRow__label"
            for="random_html_id"
          >
            <div
              style="line-height: 8px;"
            >
              <p>
                Description 
                <span
                  class="optional"
                >
                  - optional
                </span>
              </p>
              <br />
              <p
                class="sublabel"
              >
                Describe the purpose of the detector.
              </p>
            </div>
          </label>
        </div>
        <div
          class="euiFormRow__fieldWrapper"
        >
          <textarea
            class="euiTextArea euiTextArea--resizeVertical"
            id="random_html_id"
            name="description"
            placeholder="Enter detector description"
            rows="3"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
